tools/lkl/lib/virtio_net_linux_fdnet.c

```diff
@@ -47,9 +47,9 @@ static int linux_fdnet_net_tx(struct lkl_netdev *nd,
 
 	do {
 		ret = writev(nd_fdnet->fd, (struct iovec *)iov, cnt);
-	} while (ret == -1 && errno == EINTR);
+	} while (ret == -1 && (errno == EINTR || errno == EAGAIN));
```

Comment:
> epoll_wait(EPOLLOUT) should unblock when fd is writable again after EAGAIN.

right.

> Do you see failure caused by EAGAIN? You test hangs?

yes.  after EAGAIN,  lkl_sys_send(2)  **sometimes** hangs.  It sometimes recovers by next tx epoll_wait btw.


it appears that this is not a macvtap specific issue but tap also.  If I configured a small sndbuf size (e.g., 100K) by `TUNSETSNDBUF` on bridged tap, it hangs.  I've never faced this on local tap (netserver and netperf run on the same node).